### DISCUSSION TITLE
Application Startup.

### DISCUSSION THREAD

**Discussion by PhilCase** [airsdk/Adobe-Runtime-Support#1217] [Q&A] [answered]
Hi,    I need some advice concerning application startup on Android please.

We have some customers reporting slow start up times.   I've mapped out the times the app takes to start and our QA department has looked at videos sent in by the customers and tested across the devices that they have available to them.

My colleague in QA asked about the black screen that appears when the application is loading before the splash screen shows.   It displays for about 1 second on a "reasonably modern" Android device with a 64 bit processor.  On older devices it can appear for longer.  (This screen doesn't appear at all on iOS devices).

I _think_ it has something to do with the Air VM that loads and runs the app but I wanted to verify that.     

Is there a way to hide the black screen that appears before our splash screen?

Are there any ways to improve the application start up time (or apparent start up time)?   

Do I need to split the .swf file into multiple .swfs - and if so is there any form of recommendations for doing so please?

**Comment by ajwfrost** [comment]
There is a bit of a delay within the bootstrap code that's loading in the ActionScript virtual machine, parsing the application descriptor file, and loading in the initial assets. The initial loading is also then dependent upon the first SWF file, that is mentioned in the descriptor, and also on any ANEs.

In case it helps, it might be worth checking the ANEs and seeing if any of the "library.swf" files that are in there have got lots of different ABC blocks (or the same for your main swf file). Normally with a release build, all the different ActionScript classes are merged into a single ABC block in the SWF, but we've seen some issues with the component compiler by default leaving these in separate blocks, and this slows down the validation and parsing of a SWF file quite a lot...

It may then be worth just having the initial assets/logic in your main SWF file and having this come up on the screen with some other splash logo/animation whilst it loads in other SWF content and uploads any graphical images in the background.

And then to help with the perceived start-up, it may be possible to add your own splash image via something like the below:
(corrected link in the post below...)

thanks

**Reply by marchbold** [reply]
FYI Latest version of that tutorial is here: https://docs.airnativeextensions.com/docs/tutorials/android-splash-screen (should be pretty much the same though)

**Comment by PhilCase** [comment]
Thanks for your advice.

This is what we see with a Samsung J5 device from 2016.

https://user-images.githubusercontent.com/11427050/135445501-6c8d4e4b-14bf-4d0e-baec-b62f5eb9efb5.MOV

How do I look for the "ABC Blocks" please?   

I'm also unsure about partitioning the app into separate swf files.   Is there any advice for that on the web please?

**Reply by ajwfrost** [reply]
Yes that's a bit of a delay :-(   

For ABC blocks, the simplest thing is to open the SWF files using "SWF Investigator". 
https://sourceforge.net/projects/swfinvestigator.adobe/

Here's a basic example - but hacking the build command so that I get all the built-in classes included in my SWF:
![image](https://user-images.githubusercontent.com/6400804/135449187-16adb2c9-d628-44d1-a16f-00fae838964c.png)

This has all of the class definitions still within one ABC block. But if I build with "optimize=false":

![image](https://user-images.githubusercontent.com/6400804/135449379-f8d9f0dc-7699-4635-ae8b-7206c3ba638c.png)

In this second one, you can see the "DoABC2 (82)" block count on the right is now 180. There's no real reason why any SWF should have more than one block of ActionScript byte code...

We had a customer who was including ANEs or linking against external SWC files or similar, and their start-up time was improved by around .6s if I recall, just by merging all of the ABC blocks into one.

If you do need to merge things and can't change the build settings to rebuild things from source: https://www.buraks.com/abcmerge/ can do this for you (although I've always been a little concerned at their note of an 'undisclosed use case' ..!)

Best practice for splitting an app into separate SWFs, I will probably leave for other folk to suggest ideas who have got more practical experience of it...

thanks

**Comment by PhilCase** [comment]
This is what I get with a release build for Android.

![image](https://user-images.githubusercontent.com/11427050/136239717-f69177ee-d0a2-4fdb-b372-f205559d39c1.png)

I'll see if I can find the build file.

**Reply by ajwfrost** [reply]
There's only two ABC blocks which is good. But 8.5MB of SWF that needs to be decompressed and then over 900 shapes and 365 embedded bitmaps - these should only be parsed on the frame that they're used, but there's only one frame in this so it's probably going to be a little slow for start-up.

For perception of performance the best thing may be to just have a very small and simple SWF provide a backdrop and 'loading' type animation whilst the rest of this is then loaded in asynchronously...

**Reply by PhilCase** [reply]
Would each control on each view be a classes as one of the Shapes?   Is a View a shape too?
If they all need loading as opposed to being dynamically created then it's a bit confusing to me.
Should I make one "base class" for similar views and create subclasses for specific functionality (for screens that include on-screen keyboards where the default soft keyboard has not been used).

The number of Bitmaps used is likely down to MultipleDPIBitmapSource.   Is there a better method for supporting different DPIs than using that mechanism please?

<s:MultiDPIBitmapSource 
		source120dpi="@Embed('/assets/dpi120/SmartCheckLogo.png')"
		source160dpi="@Embed('/assets/dpi160/SmartCheckLogo.png')"
		source240dpi="@Embed('/assets/dpi240/SmartCheckLogo.png')"
		source320dpi="@Embed('/assets/dpi320/SmartCheckLogo.png')"
		source480dpi="@Embed('/assets/dpi480/SmartCheckLogo.png')"/>
</s:source>


I've not used separate swfs yet.    So I don't fully appreciate the "best practice" for their use?    Do you have any pointers please?
Can multiple dpi support be done using one swf per icon set?

**Reply by ajwfrost** [reply]
Ah - so looking again, and having had it pointed out that this is Flex, there are actually two frames in there. So, traditionally this would mean the first frame is where Flex puts it system loader and displays a progress bar whilst loading in all the assets needed, and then frame 2 is where the full application logic is displayed. This is approximately equivalent to having one swf to display an image whilst a second swf is loaded in... although it's still just one SWF. Not sure whether you see any benefit if you don't use compression on the SWF itself?

But if you're using Flex then yes it's going to be a little slow.. Flex is just a fairly heavyweight set of components and classes. That idea of a multi-bitmap source is a case in point, but doing something that's more optimised would mean having to do something in a platform-specific mechanism (like using resource files in different dpi folders on Android) which negates the benefit of Flex.

There are some points in the below PDF that are quite useful, plus a section on splash screens. I would also recommend using Adobe Scout to see whether this gives any hints as to whether there are particular aspects of the code that are taking an excessive amount of time that could be trimmed somehow..
https://help.adobe.com/archive/en_US/flex/mobileapps/developing_mobile_apps_flex_4.6.pdf

thanks

**Comment by mobile-solutions-ltd** [comment]
We've done additional work into this.   I've made multiple versions of the app included and excluding libraries through the use of custom application manifests and CONFIG::xxxxx statements to include blocks of code.  (I'm not sure if conditional compilation done this way will help or hinder us.)

We've found that App starts much more quickly without distriqt's push notification support.    I'm not sure if this indicates a problem with the push notification configuration on the device or not.    Excluding push notification support drops the size of our APK by about 17MB.
I'm questioning whether I should delay setting it up until the home screen of the app has been displayed.

We're still working out whether this helps our user base.

In addition we are exploring whether the additional GPS monitoring routines in our later releases could be the cause of the problem.     I've noted that they can start whilst the splash screen is showing and that "bad code" can cause the boot to trip up and crash out of the application.

I've still done nothing regarding splitting the SWFs.   I'm not even sure where to start with that one.

I didn't get very far with Adobe Scout.     I've also tried SWFInvestigator but it stopped working.

**Comment by PhilCase** [comment, marked as answer]
Thanks for all the help with this.    

The problem was in start-up with a file being read in to set up the application.   It was becoming corrupted.  

We've fixed the issue and the Google Play Stats have settled back down again.
